src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateActions = [];
+  const stateCopy = {
+    ...state,
+  };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clearProperties(stateCopy);
+        break;
+
+      default:
+        return `Error, unknown action type: ${action.type}`;
+    }
+    stateActions.push({ ...stateCopy });
```

Comment:
[CODE STYLE]: It's better to push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after the current action is processed. This is already correctly implemented, but ensure that the logic within your case blocks doesn't accidentally modify the state before it's pushed.